assert/assertions.go

```diff
@@ -2052,23 +2060,83 @@ func EventuallyWithT(t TestingT, condition func(collect *CollectT), waitFor time
 	}
 }
 
+// EventuallySync asserts that a given condition will be met before waitFor time
+// plus the time taken for the last call to condition to return. The condition
+// is considered met when the condition function does not report failures on the
+// CollectT passed to it. The supplied CollectT collects all errors from one
+// tick (if there are any) and if the condition is not met before EventuallySync
+// returns, then the collected errors of the last tick are copied to t. The
+// condition function is called synchronously immediately and then every tick
+// duration after that. EventuallySync will adjust the time interval or drop
+// ticks to make up for slow condition functions.
+//
+//	externalValue := false
+//	go func() {
+//		time.Sleep(8*time.Second)
+//		externalValue = true
+//	}()
+//	assert.EventuallySync(t, func(c *assert.CollectT) {
+//		// add assertions as needed; any assertion failure will fail the current tick
+//		assert.True(c, externalValue, "expected 'externalValue' to be true")
+//	}, 10*time.Second, 1*time.Second, "external state has not changed to 'true'; still false")
+func EventuallySync(t TestingT, condition func(*CollectT), waitFor, tick time.Duration, msgAndArgs ...interface{}) bool {
+	deadline := time.Now().Add(waitFor)
+
+	tickerCh := make(chan time.Time)
+	close(tickerCh)
+	ticker := (<-chan time.Time)(tickerCh)
+	if tick > 0 {
+		timeTicker := time.NewTicker(tick)
+		defer timeTicker.Stop()
+		ticker = timeTicker.C
+	}
+
+	for {
+		collect := new(CollectT)
+
+		wait := make(chan struct{})
+		go func() {
```

Comment:
`CollectT.FailNow` will call `runtime.GoExit`. If we didn't use a new goroutine for each callback then the user calling `FailNow` would stop the whole test. This is the only reason we can't do this without any additional goroutines at all.

Reply:
It looks like @dolmen [doesn't like `CollectT`](https://github.com/stretchr/testify/pull/1606#issuecomment-2165668478), and I can't say I disagree with him.

So we actually could make this work with no additional goroutines. The choice is to either use `func() bool` or use an interface with an un-exported implementation. I find that every time I use Eventually I want to use assertions inside the condition function, so the latter is my preference. But we can panic an un-exported value rather than calling `runtime.Goexit`. All panics from the condition function should be handled regardless.

Reply:
I'm not sure I totally follow. It sounds like an unexported implementation makes sense, but could you sketch out what that code would look like?